### PR TITLE
Feature/issue 2752 update rhat and ess (part II)

### PR DESCRIPTION
**Pull request by roualdes**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

Re issue https://github.com/stan-dev/stan/issues/2752: offer interfaces a unified calculation of rhat, and add checks in both calculations of ESS and Rhat to better match the reference implementation in R at https://github.com/stan-dev/rstan/blob/develop/rstan/rstan/R/monitor.R.  

Re Analysis API (see https://github.com/stan-dev/stan/issues/2569 and/or https://github.com/stan-dev/stan/pull/2575) this PR moves potential_scale_reduction to analyze service.  To be more consistent with the discussions in issue https://github.com/stan-dev/stan/issues/2752, the splitting of split_potential_scale_reduction now follows the design of the splitting of the effective sample size calculations.

#### Intended Effect

Provide a unified calculation of split Rhat.

#### How to Verify

Tests added in `src/test/unit/analyze/mcmc`.

#### Side Effects

None intended.

#### Documentation

Inline.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): California State University, Chico

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by betanalpha** [comment]
I’m a bit slammed but I’ll try to review by the end of next week!

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp [hunk 1]
```diff
@@ -35,6 +37,30 @@ namespace analyze {
       num_draws = std::min(num_draws, sizes[chain]);
     }
 
+    if ( num_draws < 3 ) {
```
Comment (betanalpha) [inline src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: This edge case should be noted in the function documentation.  Perhaps also a note here about why less than three (e.g. variance cannot be estimated for less than three draws).

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp [hunk 2]
```diff
@@ -35,6 +37,30 @@ namespace analyze {
...
+      return std::numeric_limits<double>::quiet_NaN();
+    }
+
+    Eigen::VectorXd draw_val = Eigen::VectorXd::Random(num_chains);
```
Comment (betanalpha) [inline src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: `draw_val` needs to be documented.  The random initialization is also dangerous as it can yield non-reproducible results -- I recommend going with (1, 2, 3, ..., num_chains) to give distinct but determinism values.

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp [hunk 3]
```diff
@@ -35,6 +37,30 @@ namespace analyze {
...
+          return std::numeric_limits<double>::quiet_NaN();
+        }
+      }
+
+      if ( draw.isApproxToConstant(draw(0)) ) {
+        draw_val(chain) = draw(0);
+      }
+    }
+
+    if ( draw_val.isApproxToConstant(draw_val(0)) ) {
```
Comment (betanalpha) [inline src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: Do we not want to return NaN here?  If everything is constant then the asymptotic variance, and hence ESS, isn't particularly well-defined.

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+   * @param draws stores pointers to arrays of chains
+   * @param sizes stores sizes of chains
+   * @return potential scale reduction for the specified parameter
+   */
+  inline
+  double compute_potential_scale_reduction(std::vector<const double*> draws,
+                                           std::vector<size_t> sizes) {
+    int num_chains = sizes.size();
+    size_t num_draws = sizes[0];
+    for ( int chain = 1; chain < num_chains; ++chain ) {
```
Comment (betanalpha) [inline src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: The function doc says the code assumes that the chains are all of equal size, but this code doesn't.  Either the function doc should change ("chains are trimmed from the back to match the length of the shortest chain") or this code should check equal size and throw a NaN or an exception if they're not the same.

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+          return std::numeric_limits<double>::quiet_NaN();
+        }
+      }
+
+      if ( draw.isApproxToConstant(draw(0)) ) {
+        draw_val(chain) = draw(0);
+      }
+    }
+
+    if ( draw_val.isApproxToConstant(draw_val(0)) ) {
```
Comment (betanalpha) [inline src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Same comment as above -- if everything is constant then do we want a NaN here instead of a "perfect" value?

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    Eigen::VectorXd chain_mean(num_chains);
+    Eigen::VectorXd chain_var(num_chains);
+
+    for ( int chain = 0; chain < num_chains; chain++ ) {
+      Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, 1>>
+        draw(draws[chain], sizes[chain]);
+      chain_mean(chain) = draw.mean();
+      chain_var(chain) = (draw.array() - chain_mean(chain)).square().sum()
```
Comment (betanalpha) [inline src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: This method of computing the empirical variance is prone to numerical inaccuracy.  We should use a Welford accumulator -- these are implemented in Boost but the code is short enough that it can be done by hand here.  See for example https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/stan/math/prim/mat/fun/welford_var_estimator.hpp.

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp (hunk 3 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: A valid point and I agree.  I was going for consistency with the original issue's reference implementation: https://github.com/stan-dev/rstan/blob/8caa8728669ae2df75167202e08dff930d4d7c64/rstan/rstan/R/monitor.R#L213

Should we over rule the consistency argument in favor of returning NaN here?

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp (hunk 5 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Same as above.  Should we over rule the consistency argument in favor of returning NaN here?

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp (hunk 6 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Nice.  

Is it too much for this PR to make this same upgrade in stan-dev/stan's [analyze/mcmc/autocovariance.hpp](https://github.com/stan-dev/stan/blob/f372f98065461b1d7b4b4fe3d1b0bd984138887a/src/stan/analyze/mcmc/autocovariance.hpp#L87) also?

Comment (betanalpha) [reply src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Could you make an issue and then put together a PR for updating this code and the code in autocovariance.hpp?  That way we can push this PR as is and contain the better variance estimation to its own PR.

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp (hunk 5 above)
Comment (betanalpha) [reply src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Let me start a Discourse thread appealing for opinion on this.  Worst case I'll approve for the consistency of the current behavior and we'll fix with a separate PR.

**Comment by betanalpha** [comment]
Sorry for the delay -- I was hoping that the Discourse thread would evolve but it seems to have come to an uneasy equilibrium.  For now I'm going to request that you return `NaN` for Rhat and ESS when the chains are constant and we can handle the management of all of the possible cases in the MCMC standard error code (for example returning SE=0 if variance is finite but ESS=NaN, returning SE=NaN if both variance and ESS are NaN, etc) and check functions (reporting that NaN means constant values and users are responsible for verifying whether that's expected or not).

Once that small change is made I'll approve.

Replacing the variance calculations with accumulators can be pushed to a new issue/PR.

Thanks for your patience!

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp (hunk 1 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: Done

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp (hunk 2 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: (1, 2, 3, ... num_chains), check.  Would you like an inline comment or for `draw_val` to be mentioned in the function documentation?

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp (hunk 4 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Made explicit the trimming.

File: src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp (hunk 6 above)
Comment (roualdes) [reply src/stan/analyze/mcmc/compute_potential_scale_reduction.hpp]: Will do.

File: src/stan/analyze/mcmc/compute_effective_sample_size.hpp (hunk 2 above)
Comment (betanalpha) [reply src/stan/analyze/mcmc/compute_effective_sample_size.hpp]: Typically inline if it doesn't affect the outcome of the function evaluation (and hence is just an implementation detail) and in the function doc if it does affect output values so that the client is aware of any expected edge cases.

**Comment by betanalpha** [comment]
Just waiting on the draw_vals doc and then we're good to go!

**Comment by roualdes** [comment]
Yesterday, I discovered an intermittent error in the tests within the file `split_chains_test.cpp`.  I believe my previous test was using incorrect logic.  Would you please give `split_chains_test.cpp` a look over?  The update should be correct now, but as always a second set of eyes rarely hurts.  Thanks.
